pyvista/core/filters/data_set.py

```diff
@@ -68,6 +68,108 @@ def _clip_with_function(
             clipped = self.extract_cells(np.unique(clipped.cell_data['cell_ids']))
         return clipped
 
+    def iterative_closest_point(
+        self,
+        target,
+        max_landmarks=100,
+        max_mean_distance=1e-5,
+        max_iterations=500,
+        check_mean_distance=True,
+        start_by_matching_centroids=True,
+        return_matrix=False,
+    ):
+        """Perform iterative closest point registration between two datasets.
+
+        Uses the VTK class `vtkIterativeClosestPointTransform
+        <https://vtk.org/doc/nightly/html/classvtkIterativeClosestPointTransform.html>`_
+
+        Parameters
+        ----------
+        target : pyvista.DataSet
+            The target dataset to align to.
+
+        max_landmarks : int, default: 100
+            The maximum number of landmarks.
+
+        max_mean_distance : float, default: 1e-5
+            The maximum mean distance for convergence.
+
+        max_iterations : int, default: 500
+            The maximum number of iterations.
+
+        check_mean_distance : bool, default: True
+            Whether to check the mean distance for convergence.
+
+        start_by_matching_centroids : bool, default: True
+            Whether to start the alignment by matching centroids. Default is True.
+
+        return_matrix : bool, default: False
+            Return the transform matrix as well as the aligned mesh.
+
+        Returns
+        -------
+        aligned : pyvista.DataSet
+            The dataset aligned to the target mesh.
+
+        matrix : numpy.ndarray
+            Transform matrix to transform the input dataset to the target dataset.
```

Comment:
The functionality seems great, but the name is super confusing. We have filters like `find_closest_point()`, which finds the closest point (:exploding_head:). This is `iterative_closest_point()` which... returns a mesh and a matrix? Can we call it something more descriptive? `iterative_closest_point_transform()` would be a lot better to me, but of course that's very verbose.

Reply:
Maybe cater to clueless users such as myself and call it something like `align_transform()`, and mention the ICP algorithm in the description so that experts looking for it can find the appropriate method?

Reply:
I'm on the same page here. PR title made me think of something totally different. I think `transform_by_closest_point()` or similar would be a good name (with `transform` first to put it next to other `transform*` methods)

Reply:
Guess it's my familarity with ICP (iterative closest point).

I'd prefer something like "align", but second place would be `transform_by_alignment`, which we could consider exposing different alignment algorithms.

Reply:
`align()` would also be a good choice for this and is in line with other methods that require a second dataset, e.g., `sample()`. Leaving this mostly to your discretion, @akaszynski

Reply:
Though considering `Transform` is in the name of the VTK algorithm, I'm leaning towards `transform_*`... perhaps even `transform_align()`?

Reply:
Align seems the most intutive to me. I'll wait for others to weigh in here before proceeding.

Reply:
Alright, going with `align` here.